forbidigo/forbidigo.go

```diff
@@ -163,44 +195,198 @@ func (v *visitor) Visit(node ast.Node) ast.Visitor {
 			return nil
 		}
 		return v
+	// The following two are handled below.
 	case *ast.SelectorExpr:
 	case *ast.Ident:
+	// Everything else isn't.
 	default:
 		return v
 	}
+
+	// The text as it appears in the source is always used because issues
+	// use that. It's used for matching unless usage of type information
+	// is enabled.
+	srcText := v.textFor(node)
+	matchText, pkgText := v.updateText(node, srcText)
 	for _, p := range v.linter.patterns {
-		if p.pattern.MatchString(v.textFor(node)) && !v.permit(node) {
+		if p.re.MatchString(matchText) &&
+			(p.Package == "" || p.pkgRe.MatchString(pkgText)) &&
+			!v.permit(node) {
 			v.issues = append(v.issues, UsedIssue{
-				identifier: v.textFor(node),
-				pattern:    p.pattern.String(),
+				identifier: srcText, // Always report the expression as it appears in the source code.
+				pattern:    p.re.String(),
 				pos:        node.Pos(),
-				position:   v.fset.Position(node.Pos()),
-				customMsg:  p.msg,
+				position:   v.runConfig.Fset.Position(node.Pos()),
+				customMsg:  p.Msg,
 			})
 		}
 	}
 	return nil
 }
 
+// textFor returns the expression as it appears in the source code (for
+// example, <importname>.<function name>).
 func (v *visitor) textFor(node ast.Node) string {
 	buf := new(bytes.Buffer)
-	if err := printer.Fprint(buf, v.fset, node); err != nil {
-		log.Fatalf("ERROR: unable to print node at %s: %s", v.fset.Position(node.Pos()), err)
+	if err := printer.Fprint(buf, v.runConfig.Fset, node); err != nil {
+		log.Fatalf("ERROR: unable to print node at %s: %s", v.runConfig.Fset.Position(node.Pos()), err)
 	}
 	return buf.String()
 }
 
+// pkgTextFor expands the selector in a selector expression to the full package
+// name and (for variables) the type:
+//
+// - example.com/some/pkg.Function
+// - example.com/some/pkg.CustomType.Method
+//
+// It updates the text to match against and fills the package string if possible,
+// otherwise it just returns.
+func (v *visitor) updateText(node ast.Node, srcText string) (matchText, pkgText string) {
+	// The text to match against is the literal source code if we cannot
+	// come up with something different.
+	matchText = srcText
+
+	if v.runConfig.TypesInfo == nil {
+		return
+	}
+
+	location := v.runConfig.Fset.Position(node.Pos())
+
+	// TODO: do type switch here instead of multiple if checks.
+	if ident, ok := node.(*ast.Ident); ok {
+		object, ok := v.runConfig.TypesInfo.Uses[ident]
+		if !ok {
+			// No information about the identifier. Should
+			// not happen, but perhaps there were compile
+			// errors?
+			v.runConfig.DebugLog("%s: unknown identifier %q", location, srcText)
+			return
+		}
+		if pkg := object.Pkg(); pkg != nil {
+			matchText = importPathBase(pkg.Path()) + "." + object.Name()
+			pkgText = pkg.Path()
+			v.runConfig.DebugLog("%s: identifier: %q -> %q in package %q", location, srcText, matchText, pkgText)
+		} else {
+			matchText = object.Name()
```

Comment:
Am I right in thinking that this should just be the same as the identifier itself?